File: cpp/threat_engine/phishing_detector.cpp
// cpp/threat_engine/phishing_detector.cpp
#include "phishing_detector.hpp"
#include <algorithm>
#include <sstream>
#include <cmath>
#include <openssl/sha.h>
#include <iomanip>

namespace amd_security {
namespace threat_detection {

PhishingDetector::PhishingDetector() {
    // Initialize suspicious domains
    suspicious_domains = {
        {"paypa1.com", 0.9f},
        {"amaz0n.com", 0.9f},
        {"go0gle.com", 0.9f},
        {"bank-verify.com", 0.95f},
        {"account-confirm.com", 0.95f},
        {"secure-login.com", 0.9f}
    };

    // Initialize suspicious TLDs
    suspicious_tlds = {".tk", ".ml", ".ga", ".cf", ".xyz", ".top"};

    // Initialize regex patterns
    try {
        suspicious_patterns.emplace_back(R"(http[s]?:\/\/\d+\.\d+\.\d+\.\d+)"); // IP addresses
        suspicious_patterns.emplace_back(R"(@)");  // @ symbol
        suspicious_patterns.emplace_back(R"(\?)");  // Question mark
    } catch (const std::regex_error& e) {
        // Log error
    }
}

PhishingDetector::~PhishingDetector() = default;

ThreatResult PhishingDetector::detect(const std::string& url, const std::string& context) {
    ThreatResult result;
    result.url_hash = sha256(url);

    // Validate URL
    if (url.empty() || !is_valid_url(url)) {
        result.is_phishing = false;
        result.confidence = 0.0f;
        result.threat_level = "safe";
        result.reasons.push_back("Invalid URL format");
        return result;
    }

    // Extract features
    URLFeatures features = extract_features(url);

    // Add context analysis
    float context_score = 0.0f;
    if (!context.empty()) {
        std::string context_lower = context;
        std::transform(context_lower.begin(), context_lower.end(), 
                      context_lower.begin(), ::tolower);
        
        const std::vector<std::string> phishing_keywords = {
            "verify", "confirm", "update", "validate", "secure"
        };
        
        int keyword_count = 0;
        for (const auto& kw : phishing_keywords) {
            if (context_lower.find(kw) != std::string::npos) {
                keyword_count++;
            }
        }
        context_score = std::min(keyword_count * 0.15f, 0.5f);
    }

    // Calculate threat score
    float threat_score = calculate_threat_score(features);
    threat_score = (threat_score * 0.8f) + (context_score * 0.2f);

    result.confidence = std::min(threat_score, 1.0f);
    result.is_phishing = threat_score >= CONFIDENCE_THRESHOLD;
    result.threat_level = result.is_phishing ? "high" : "safe";
    result.reasons = extract_reasons(features);

    return result;
}

std::vector<ThreatResult> PhishingDetector::detect_batch(const std::vector<std::string>& urls) {
    std::vector<ThreatResult> results;
    results.reserve(urls.size());

    // Use OpenMP for parallel processing (SIMD optimization)
    #pragma omp parallel for
    for (size_t i = 0; i < urls.size(); ++i) {
        results[i] = detect(urls[i]);
    }

    return results;
}

PhishingDetector::URLFeatures PhishingDetector::extract_features(const std::string& url) {
    URLFeatures features;
    features.url_length_score = score_url_length(url);
    features.domain_score = score_domain(url);
    features.special_char_score = score_special_chars(url);
    features.ip_address_score = score_ip_address(url);
    features.subdomain_score = score_subdomains(url);
    return features;
}

float PhishingDetector::score_url_length(const std::string& url) {
    size_t length = url.length();
    if (length > 200) return 0.8f;
    if (length > 100) return 0.5f;
    return 0.0f;
}

float PhishingDetector::score_domain(const std::string& url) {
    try {
        size_t start = url.find("://") + 3;
        size_t end = url.find('/', start);
        if (end == std::string::npos) end = url.length();
        
        std::string domain = url.substr(start, end - start);
        std::transform(domain.begin(), domain.end(), domain.begin(), ::tolower);

        // Check suspicious domains
        auto it = suspicious_domains.find(domain);
        if (it != suspicious_domains.end()) {
            return it->second;
        }

        // Check suspicious TLDs
        for (const auto& tld : suspicious_tlds) {
            if (domain.find(tld) != std::string::npos) {
                return 0.7f;
            }
        }

        // Check lookalike domains
        if (is_lookalike_domain(domain)) {
            return 0.7f;
        }

        return 0.0f;
    } catch (...) {
        return 0.0f;
    }
}

float PhishingDetector::score_special_chars(const std::string& url) {
    int special_count = std::count(url.begin(), url.end(), '@') +
                        std::count(url.begin(), url.end(), '?');
    if (special_count > 2) return 0.6f;
    return 0.0f;
}

float PhishingDetector::score_ip_address(const std::string& url) {
    // Check for IP address pattern
    std::regex ip_pattern(R"(http[s]?://\d+\.\d+\.\d+\.\d+)");
    return std::regex_search(url, ip_pattern) ? 0.8f : 0.0f;
}

float PhishingDetector::score_subdomains(const std::string& url) {
    try {
        size_t start = url.find("://") + 3;
        size_t end = url.find('/', start);
        if (end == std::string::npos) end = url.length();
        
        std::string domain = url.substr(start, end - start);
        int subdomain_count = std::count(domain.begin(), domain.end(), '.');
        
        if (subdomain_count > 3) return 0.6f;
        return 0.0f;
    } catch (...) {
        return 0.0f;
    }
}

float PhishingDetector::calculate_threat_score(const URLFeatures& features) {
    float total = features.url_length_score + 
                  features.domain_score + 
                  features.special_char_score + 
                  features.ip_address_score + 
                  features.subdomain_score;
    
    return total / 5.0f;
}

bool PhishingDetector::is_lookalike_domain(const std::string& domain) {
    const std::map<std::string, std::vector<std::string>> lookalikes = {
        {"paypa", {"paypal"}},
        {"amaz", {"amazon"}},
        {"goog", {"google"}},
        {"face", {"facebook"}}
    };

    for (const auto& [key, similar] : lookalikes) {
        if (domain.find(key) != std::string::npos) {
            for (const auto& sim : similar) {
                if (domain.find(sim) != std::string::npos && 
                    domain != sim + ".com") {
                    return true;
                }
            }
        }
    }
    return false;
}

bool PhishingDetector::is_valid_url(const std::string& url) {
    std::regex url_regex(R"(http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),])+)");
    return std::regex_search(url, url_regex);
}

std::vector<std::string> PhishingDetector::extract_reasons(const URLFeatures& features) {
    std::vector<std::string> reasons;
    
    if (features.url_length_score > 0.5f) {
        reasons.push_back("Unusually long URL");
    }
    if (features.domain_score > 0.5f) {
        reasons.push_back("Suspicious domain name");
    }
    if (features.special_char_score > 0.5f) {
        reasons.push_back("Suspicious special characters");
    }
    if (features.ip_address_score > 0.5f) {
        reasons.push_back("Using IP address instead of domain");
    }
    if (features.subdomain_score > 0.5f) {
        reasons.push_back("Too many subdomains");
    }
    
    if (reasons.empty()) {
        reasons.push_back("URL appears legitimate");
    }
    
    return reasons;
}

std::string PhishingDetector::sha256(const std::string& input) {
    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256(reinterpret_cast<const unsigned char*>(input.c_str()), 
           input.length(), hash);
    
    std::stringstream ss;
    for (int i = 0; i < 8; ++i) {  // First 8 bytes for short hash
        ss << std::hex << std::setw(2) << std::setfill('0') 
           << static_cast<int>(hash[i]);
    }
    return ss.str();
}

} // namespace threat_detection
} // namespace amd_security

File: src/threat_detection/phishing_detector.py
# src/threat_detection/phishing_detector.py
"""
Phishing Link Detection Module
Uses lightweight ML models for real-time phishing detection
"""

import json
import re
import hashlib
from typing import Dict, Tuple, List
from config.settings import DETECTION_CONFIG, DATA_DIR
from config.logger import SecurityLogger

logger = SecurityLogger.get_logger(__name__)

class PhishingDetector:
    """Detects phishing links using heuristics and ML models"""
    
    def __init__(self):
        """Initialize phishing detector with known patterns"""
        self.confidence_threshold = DETECTION_CONFIG['phishing']['confidence_threshold']
        self.patterns_db = self._load_phishing_patterns()
        self.suspicious_domains = self._init_suspicious_domains()
        self.url_regex = re.compile(
            r'http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\\(\\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+'
        )
        logger.info("PhishingDetector initialized")
    
    def _load_phishing_patterns(self) -> Dict:
        """Load phishing patterns from database"""
        try:
            patterns_file = DATA_DIR / "phishing_patterns.json"
            if patterns_file.exists():
                with open(patterns_file, 'r') as f:
                    return json.load(f)
        except Exception as e:
            logger.error(f"Error loading phishing patterns: {e}")
        
        # Default patterns if file doesn't exist
        return {
            "keywords": ["verify", "confirm", "update", "validate", "secure"],
            "suspicious_tlds": [".tk", ".ml", ".ga", ".cf"],
        }
    
    def _init_suspicious_domains(self) -> set:
        """Initialize list of known suspicious domains"""
        return {
            "paypa1.com", "amaz0n.com", "go0gle.com", "bank-verify.com",
            "account-confirm.com", "secure-login.com", "update-verify.com"
        }
    
    def detect(self, url: str, context: str = "") -> Dict:
        """
        Detect if a URL is phishing
        
        Args:
            url: URL to check
            context: Additional context (email body, page text, etc.)
        
        Returns:
            Dict with threat assessment
        """
        try:
            # Validate URL format
            if not url or not self._is_valid_url(url):
                return {
                    "is_phishing": False,
                    "confidence": 0.0,
                    "threat_level": "safe",
                    "reasons": ["Invalid URL format"]
                }
            
            # Heuristic-based detection
            features = self._extract_url_features(url)
            features['context_score'] = self._analyze_context(context)
            
            # Calculate threat score
            threat_score = self._calculate_threat_score(features)
            
            # Determine if phishing
            is_phishing = threat_score >= self.confidence_threshold
            
            result = {
                "is_phishing": is_phishing,
                "confidence": min(threat_score, 1.0),
                "threat_level": "high" if is_phishing else "safe",
                "reasons": self._generate_reasons(features),
                "url_hash": hashlib.sha256(url.encode()).hexdigest()[:8]
            }
            
            logger.info(f"Phishing detection - URL: {url[:50]}... - Score: {threat_score:.2f}")
            return result
        
        except Exception as e:
            logger.error(f"Error in phishing detection: {e}")
            return {
                "is_phishing": False,
                "confidence": 0.0,
                "threat_level": "unknown",
                "reasons": ["Error in detection"]
            }
    
    def _is_valid_url(self, url: str) -> bool:
        """Check if URL has valid format"""
        return bool(self.url_regex.match(url))
    
    def _extract_url_features(self, url: str) -> Dict:
        """Extract features from URL for threat scoring"""
        features = {
            "url_length_score": self._score_url_length(url),
            "domain_score": self._score_domain(url),
            "special_char_score": self._score_special_chars(url),
            "ip_address_score": self._score_ip_address(url),
            "subdomain_score": self._score_subdomains(url),
        }
        return features
    
    def _score_url_length(self, url: str) -> float:
        """Score based on URL length (very long URLs are suspicious)"""
        length = len(url)
        if length > 200:
            return 0.8
        elif length > 100:
            return 0.5
        return 0.0
    
    def _score_domain(self, url: str) -> float:
        """Score domain for suspicious characteristics"""
        try:
            # Extract domain
            domain = url.split('/')[2].lower()
            
            # Check if in suspicious list
            if domain in self.suspicious_domains:
                return 0.9
            
            # Check for suspicious TLDs
            for tld in self.patterns_db.get("suspicious_tlds", []):
                if domain.endswith(tld):
                    return 0.7
            
            # Check for look-alike domains
            if self._is_lookalike_domain(domain):
                return 0.7
            
            return 0.0
        except Exception as e:
            logger.warning(f"Error scoring domain: {e}")
            return 0.0
    
    def _is_lookalike_domain(self, domain: str) -> bool:
        """Check if domain looks like a legitimate site"""
        lookalikes = {
            "paypa": ["paypal"],
            "amaz": ["amazon"],
            "goog": ["google"],
            "face": ["facebook"],
        }
        
        for key, similar in lookalikes.items():
            if key in domain:
                for sim in similar:
                    if sim in domain and domain != f"{sim}.com":
                        return True
        return False
    
    def _score_special_chars(self, url: str) -> float:
        """Score based on suspicious special characters"""
        special_chars = url.count('@') + url.count('?')
        if special_chars > 2:
            return 0.6
        return 0.0
    
    def _score_ip_address(self, url: str) -> bool:
        """Check if URL uses IP address instead of domain"""
        ip_pattern = re.compile(r'http[s]?://\d+\.\d+\.\d+\.\d+')
        if ip_pattern.match(url):
            return 0.8
        return 0.0
    
    def _score_subdomains(self, url: str) -> float:
        """Score based on number of subdomains"""
        domain_part = url.split('/')[2]
        subdomain_count = domain_part.count('.')
        if subdomain_count > 3:
            return 0.6
        return 0.0
    
    def _analyze_context(self, context: str) -> float:
        """Analyze surrounding context for phishing indicators"""
        if not context:
            return 0.0
        
        context_lower = context.lower()
        phishing_keywords = self.patterns_db.get("keywords", [])
        
        keyword_count = sum(1 for kw in phishing_keywords if kw in context_lower)
        return min(keyword_count * 0.15, 0.5)
    
    def _calculate_threat_score(self, features: Dict) -> float:
        """Calculate overall threat score from features"""
        scores = list(features.values())
        if not scores:
            return 0.0
        return sum(scores) / len(scores)
    
    def _generate_reasons(self, features: Dict) -> List[str]:
        """Generate human-readable reasons for threat assessment"""
        reasons = []
        
        if features['url_length_score'] > 0.5:
            reasons.append("Unusually long URL")
        if features['domain_score'] > 0.5:
            reasons.append("Suspicious domain name")
        if features['special_char_score'] > 0.5:
            reasons.append("Suspicious special characters in URL")
        if features['ip_address_score'] > 0.5:
            reasons.append("Using IP address instead of domain")
        if features['subdomain_score'] > 0.5:
            reasons.append("Too many subdomains")
        if features['context_score'] > 0.3:
            reasons.append("Context contains phishing keywords")
        
        if not reasons:
            reasons.append("URL appears legitimate")
        
        return reasons

# Demo usage
if __name__ == "__main__":
    detector = PhishingDetector()
    
    test_urls = [
        "https://www.google.com",
        "https://paypa1.com/login",
        "https://192.168.1.1/admin",
        "http://bank-verify.com/update-account?verify=true&confirm=secure",
    ]
    
    for url in test_urls:
        result = detector.detect(url, "Please verify your account")
        print(f"\nURL: {url}")
        print(f"  Phishing: {result['is_phishing']}")
        print(f"  Confidence: {result['confidence']:.2f}")
        print(f"  Reasons: {result['reasons']}")
